CORDA-3850 Add a per flow lock

lankydan opened the pull request:
Add a lock to `StateMachineState` that is used when interacting with a
flow. This prevents multiple threads from interacting with a flow at the
same time, therefore stopping race conditions and weird timing issues.

To take the lock call `FlowStateMachineImpl.withFlowLock`.

The lock should be taken at all points where interaction with a flow
starts. Currently this includes starting a transition
(`FlowStateMachineImpl.processEvent`) and kill flow (`SMM.killFlow`).

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -245,36 +245,28 @@ internal class SingleThreadedStateMachineManager(
-                return@withLock if (state != null) {
-                    state.value.isKilled = true
-                    flow.fiber.scheduleEvent(Event.DoRemainingWork)
+                flow.fiber.withFlowLock {
```
Should this double check that the flow is still valid once the lock has been acquired?

e.g. is `flows[id] == flow` ?  Or maybe check if `flow.isKilled` already

lankydan replied:
yes, that makes sense.

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt:
```diff
@@ -135,27 +136,37 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
 
+    internal inline fun <RESULT> withFlowLock(block: FlowStateMachineImpl<R>.() -> RESULT): RESULT {
+        return transientState!!.value.lock.withLock { block(this) }
+//        return block(this)
```
Remove this before this gets merged.

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -245,36 +245,28 @@ internal class SingleThreadedStateMachineManager(
         val killFlowResult = innerState.withLock {
```
So this could (I think) be restructured to only take this lock when we grab the flow now. This probably only matters in ENT.

lankydan committed:
CORDA-3850 Add a per flow lock

Add a lock to `StateMachineState`, allowing every flow to lock
themselves when performing a transition or when an external thread (such
as `killFlow`) tries to interact with a flow from occurring at the same
time.

Doing this prevents race-conditions where the external threads mutate
the database or the flow's state causing an in-flight transition to
fail.

A `Semaphore` is used to acquire and release the lock. A `ReentrantLock`
is not used as it is possible for a flow to suspend while locked, and
resume on a different thread. This causes a `ReentrantLock` to fail when
releasing the lock because the thread doing so is not the thread holding
the lock. `Semaphore`s can be used across threads, therefore bypassing
this issue.

The lock is copied across when a flow is retried. This is to prevent
another thread from interacting with a flow just after it has been
retried. Without copying the lock, the external thread would acquire the
old lock and execute, while the fiber thread acquires the new lock and
also executes.
